src/stan/mcmc/hmc/static_uniform/base_static_uniform.hpp

```diff
@@ -0,0 +1,159 @@
+#ifndef STAN_MCMC_HMC_UNIFORM_BASE_STATIC_UNIFORM_HPP
+#define STAN_MCMC_HMC_UNIFORM_BASE_STATIC_UNIFORM_HPP
+
+#include <stan/interface_callbacks/writer/base_writer.hpp>
+#include <stan/mcmc/hmc/base_hmc.hpp>
+#include <stan/mcmc/hmc/hamiltonians/ps_point.hpp>
+#include <boost/math/special_functions/fpclassify.hpp>
+#include <boost/random/uniform_int_distribution.hpp>
+#include <cmath>
+#include <limits>
+#include <string>
+#include <vector>
+
+namespace stan {
+
+  namespace mcmc {
+    /**
+      * Hamiltonian Monte Carlo implemetnation that uniformly samples
+      * from trajectories with a static integration time
+    */
+    template <class Model,
+              template<class, class> class Hamiltonian,
+              template<class> class Integrator,
+              class BaseRNG>
+    class base_static_uniform:
+      public base_hmc<Model, Hamiltonian, Integrator, BaseRNG> {
+    public:
+      base_static_uniform(Model &m, BaseRNG& rng)
+        : base_hmc<Model, Hamiltonian, Integrator, BaseRNG>(m, rng),
+          T_(1), energy_(0) {
+        update_L_();
+      }
+
+      ~base_static_uniform() {}
+
+      sample transition(sample& init_sample,
+                        interface_callbacks::writer::base_writer& writer) {
+        this->sample_stepsize();
+
+        this->seed(init_sample.cont_params());
+
+        this->hamiltonian_.sample_p(this->z_, this->rand_int_);
+        this->hamiltonian_.init(this->z_, writer);
+
+        ps_point z_init(this->z_);
+        double H0 = this->hamiltonian_.H(this->z_);
+
+        ps_point z_sample(this->z_);
+        double sum_prob = 1;
+        double sum_metro_prob = 1;
+
+        boost::random::uniform_int_distribution<> uniform(0, L_ - 1);
+        int Lp = uniform(this->rand_int_);
+
+        for (int l = 0; l < Lp; ++l) {
+          this->integrator_.evolve(this->z_, this->hamiltonian_,
+                                   -this->epsilon_, writer);
+
+          double h = this->hamiltonian_.H(this->z_);
+          if (boost::math::isnan(h))
+            h = std::numeric_limits<double>::infinity();
+
+          double prob = std::exp(H0 - h);
+          sum_prob += prob;
+          sum_metro_prob += prob > 1 ? 1 : prob;
+
+          if (this->rand_uniform_() < prob / sum_prob)
+            z_sample = this->z_;
+        }
+
+        this->z_.ps_point::operator=(z_init);
+
+        for (int l = 0; l < L_ - 1 - Lp; ++l) {
+          this->integrator_.evolve(this->z_, this->hamiltonian_,
+                                   this->epsilon_, writer);
+
+          double h = this->hamiltonian_.H(this->z_);
+          if (boost::math::isnan(h))
+            h = std::numeric_limits<double>::infinity();
```

Comment:
Perhaps, but nothing I’d feel comfortable with at the moment.

On Apr 11, 2016, at 5:38 PM, Bob Carpenter <notifications@github.com> wrote:

> In src/stan/mcmc/hmc/static_uniform/base_static_uniform.hpp:
> 
> > +          sum_prob += prob;
> > +          sum_metro_prob += prob > 1 ? 1 : prob;
> > +
> > +          if (this->rand_uniform_() < prob / sum_prob)
> > +            z_sample = this->z_;
> > +        }
> > +
> > +        this->z_.ps_point::operator=(z_init);
> > +
> > +        for (int l = 0; l < L_ - 1 - Lp; ++l) {
> > +          this->integrator_.evolve(this->z_, this->hamiltonian_,
> > +                                   this->epsilon_, writer);
> > +
> > +          double h = this->hamiltonian_.H(this->z_);
> > +          if (boost::math::isnan(h))
> > +            h = std::numeric_limits<double>::infinity();
> Can we predict more of what's going to happen downstream once h becomes infinite? This is more a question than a request for a code change.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
>